File: prototyp/python/address.py
import sys
from python.storage import log_new_address

import utils

# ==============================================================================
# CONFIGURATION
# ==============================================================================
DEBUG = True               # True = Dummy-Interface (ipwrap0) | False = Live-Betrieb (Default-IF)
ANZAHL_IPS = 5             # Wie viele IPs generiert werden sollen
LEBENSDAUER_SEKUNDEN = 60  # TTL in Sekunden (Setze auf None für permanente IPs)

# ==============================================================================
# CORE LOGIC
# ==============================================================================

def build_address(count: int, prefix: str, before: set[str], iface: str, ttl: int = None) -> list[str]:
    created = []
    for _ in range(count):
        new_addr = utils.generate_address(prefix)
        while new_addr in before:
            new_addr = utils.generate_address(prefix)
        if utils.add_address(iface, new_addr, ttl=ttl):
            created.append(new_addr)  # ← neu
            before.add(new_addr)
            ttl_info = f" (TTL: {ttl}s)" if ttl else " (Permanent)"
            print(f"[✓] {new_addr}{ttl_info}")
            log_new_address(address=new_addr, iface=iface, ttl=ttl, tag="dev-test")
        else:
            print(f"[✗] Fehler bei {new_addr}")
    return created  



# ==============================================================================
# ADDRESS ENTRY POINT
# ==============================================================================

if __name__ == "__main__":
    utils.isroot()   

    if DEBUG:
        print("[MODUS] DEBUG-Modus aktiv. Nutze Dummy-Interface.")
        interface = utils.ensure_dummy_iface("ipwrap0")
    else:
        print("[MODUS] LIVE-Modus aktiv. Ermittle Standard-Interface...")
        interface = utils.get_default_iface()
        print(f"[i] Standard-Interface gefunden: {interface}")

    try:
        prefix = utils.get_prefix(interface)
    except SystemExit:
        if DEBUG:
            prefix = "2001:db8:1234:5678:"
            print(f"[i] Dummy hat kein Präfix. Verwende Test-Präfix: {prefix}")
        else:
            sys.exit(1)

    already_assigned = utils.current_addresses(interface)

    print(f"\nStarte Erstellung von {ANZAHL_IPS} Adressen auf '{interface}'...")
    build_address(
        count=ANZAHL_IPS, 
        prefix=prefix, 
        before=already_assigned, 
        iface=interface, 
        ttl=LEBENSDAUER_SEKUNDEN
    )

File: src/address.zig
const std = @import("std");
const utils = @import("utils.zig");

const DEBUG = true;
const ANZAHL_IPS: usize = 1;
const LEBENSDAUER_SEKUNDEN: ?u64 = 60;

fn buildAddress(gpa: std.mem.Allocator, io: std.Io, count: usize, prefix: []const u8, before: *std.StringHashMap(void), iface: []const u8, ttl: ?u64, init: std.process.Init) !std.ArrayListUnmanaged([]const u8) {
    var created: std.ArrayListUnmanaged([]const u8) = .empty;

    for (0..count) |_| {
        var new_addr = try utils.generateAddress(gpa, prefix, init);

        while (before.contains(new_addr)) {
            gpa.free(new_addr);
            new_addr = try utils.generateAddress(gpa, prefix, init);
        }

        const success = try utils.addAddress(gpa, io, iface, new_addr, ttl);
        if (success) {
            try created.append(gpa, new_addr);
            try before.put(new_addr, {});
            if (ttl) |t| {
                std.debug.print("[✓] {s} (TTL: {d}s)\n", .{ new_addr, t });
            } else {
                std.debug.print("[✓] {s} (Permanent)\n", .{new_addr});
            }
        } else {
            std.debug.print("[✗] Error at {s}\n", .{new_addr});
            gpa.free(new_addr);
        }
    }
    return created;
}

pub fn main(init: std.process.Init) !void {
    const gpa = init.gpa;
    const io = init.io;

    utils.isRoot();

    var interface: []u8 = undefined;
    if (DEBUG) {
        std.debug.print("[MODE] DEBUG mode active. Using dummy interface.\n", .{});
        interface = try utils.ensureDummyIface(gpa, io, "ipwrap0");
    } else {
        std.debug.print("[MODE] LIVE mode active. Determining default interface...\n", .{});
        interface = try utils.getDefaultIface(gpa, io);
        std.debug.print("[i] Default interface found: {s}\n", .{interface});
    }
    defer gpa.free(interface);

    var prefix: []u8 = undefined;
    if (utils.getPrefix(gpa, io, interface)) |p| {
        prefix = p;
    } else |_| {
        if (DEBUG) {
            prefix = try gpa.dupe(u8, "2001:db8:1234:5678:");
            std.debug.print("[i] Dummy has no prefix. Using test prefix: {s}\n", .{prefix});
        } else {
            std.process.exit(1);
        }
    }
    defer gpa.free(prefix);

    var already_assigned = try utils.currentAddresses(gpa, io, interface);
    defer {
        var it = already_assigned.keyIterator();
        while (it.next()) |key| gpa.free(key.*);
        already_assigned.deinit();
    }

    std.debug.print("\nStarting generation of {d} addresses on '{s}'...\n", .{ ANZAHL_IPS, interface });

    var created = try buildAddress(gpa, io, ANZAHL_IPS, prefix, &already_assigned, interface, LEBENSDAUER_SEKUNDEN, init);
    defer created.deinit(gpa);
}
